starlette_cramjam/middleware.py

```diff
@@ -1,12 +1,55 @@
 """starlette_cramjam.middleware."""
 
 import re
-from typing import Any, Optional, Set
+from typing import Any, List, Optional, Set
 
-import cramjam
 from starlette.datastructures import Headers, MutableHeaders
 from starlette.types import ASGIApp, Message, Receive, Scope, Send
 
+from starlette_cramjam.compression import Compression
+
+ACCEPT_ENCODING_PATTERN = r"^(?P<compression>\w+)(;q=(?P<qvalue>[0-1]*([.][0-9]+)?))?"
+
+DEFAULT_BACKENDS = [
+    Compression.gzip,
+    Compression.deflate,
+    Compression.br,
+]
+
+
+def get_compression_backend(
+    accepted_encoding: str, compression: List[Compression]
+) -> Optional[Compression]:
+    """Return Compression backend based on default compression and accepted preference.
+
+    Links:
+    - https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html
+    - https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Encoding
+
+    """
+    # Parse Accepted-Encoding value `gzip, deflate;q=0.1`
+    encoding_values = []
+    for encoding in accepted_encoding.replace(" ", "").split(","):
+        matched = re.match(ACCEPT_ENCODING_PATTERN, encoding)
+        if matched:
+            name, q = matched.groupdict().values()
+            quality = float(q) if q else 1.0
+            encoding_values.append((name, quality))
+
+    # Create Preference matrix
+    encoding_preference = {
+        v: [n for (n, q) in encoding_values if q == v]
+        for v in sorted({q for _, q in encoding_values}, reverse=True)
+    }
+
+    # Loop through available compression and encoding preference
+    for _, pref in encoding_preference.items():
+        for backend in compression:
+            if backend.name in pref:
+                return backend
+
+    return None
```

Comment:
maybe https://github.com/PierreQuentel/httpcompressionserver/blob/master/httpcompressionserver.py#L246-L275 is better